### ISSUE TITLE
Carth's Romance-Specific Dialogue After the Leviathan

### ISSUE THREAD

**Issue by Leilukin** [KOTORCommunityPatches/K1_Community_Patch#663]
The dialogue file for the party conversation on the Ebon Hawk immediately after the Leviathan, `ebo_carth.dlg` from the `ebo_m40ad` module, shows that Carth is meant to have dialogue that are specific to players who have his romance active and has progressed his romance subplot far enough. The script that determines the condition to trigger the romance-specific dialogue is `k_pebo_carthrom`, containing the following code:
```
int StartingConditional() {
	int int1 = (GetGlobalNumber("K_SWG_CARTH") == 13);
	return int1;
}
```

The issue is that the script makes the romance dialogue trigger only if the player has  progressed in his romance subplot at a very specific point. Since the global number for completing Carth's romance is 15, the vanilla script code means any player who has progressed the global number of Carth's romance subplot past 13 would miss out the romance-specific dialogue with Carth immediately after Leviathan.

The solution is not as simple as changing the code to `int int1 = (GetGlobalNumber("K_SWG_CARTH") >= 13)`, because the global for Carth's personal quest line, `K_SWG_CARTH` has a global number of 99 that is meant for players who completed Carth's personal quest line without his romance being active.

I have tried to fix the issue by changing the code of `k_pebo_carthrom` to the following, and has tested it out in the game:
```
int StartingConditional(){
    int iResult = GetGlobalNumber("K_SWG_CARTH");

    if ( (iResult == 13) || (iResult == 14) || (iResult == 15) )
	{
      return TRUE;
    }
    return FALSE;
}
```

The same module also has a `k_pebo_carthromn` script to check if Carth's romance has not progressed far enough to trigger the post romance-specific dialogue. This is the code for the script in vanilla:
```
int StartingConditional() {
	int int1 = (GetGlobalNumber("K_SWG_CARTH") != 13);
	return int1;
}
```

Here is my own fix for the script:
```
int StartingConditional(){
    int iResult = GetGlobalNumber("K_SWG_CARTH");

    if ( (iResult != 13) && (iResult != 14) && (iResult != 15) )
	{
      return TRUE;
    }
    return FALSE;
}
```

The K1CP team is welcome to include my version of the scripts in the Community Patch.

EDIT: I just realized that K1CP has modified the scripts `k_pebo_carthrom` and `k_pebo_carthromn` to force the player to talk to HK-47 during the post-Leviathan party scene on the Ebon Hawk, so you may want to find a way to combine the HK fix and my fix, should you decide to include my fix in the Community Patch.

**Comment by Salk73** [comment]
We can easily update JC's scripts:

```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	k_pebo_carthrom
	
	Carth romance check for post-Leviathan scene.
	
	The conditions have been changed so the player will be forced to speak to
	HK-47 during this scene (if they have recruited him) so his storyline isn't
	left hanging. Additionally, the original script was not properly checking the
  state of the romance.
	
	Returns true if the player has romanced Carth and they've talked to HK.
	
	Issue #3: 
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/3

  Issue #663:
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/663
	
	JC 2019-02-02 / Leilukin 2022-08-05                                             */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

	int nRomance;
  int nGlobal = GetGlobalNumber("K_SWG_CARTH");
	if (nGlobal > 12 && nGlobal < 16) nRomance = 1;
	else nRomance = 0;
	int nHKCheck = GetGlobalBoolean("Ebo_HK47");
	int nReturn;
	if (IsAvailableCreature(NPC_HK_47)) nReturn = nRomance * nHKCheck;
	else nReturn = nRomance;

	return nReturn;
}
```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	k_pebo_carthromn
	
	Carth romance check for post-Leviathan scene.
	
	The conditions have been changed so the player will be forced to speak to
	HK-47 during this scene (if they have recruited him) so his storyline isn't
	left hanging. Additionally, the original script was not properly checking the
  state of the romance.
	
	Returns true if the player has NOT romanced Carth and they've talked to HK.
	
	Issue #3: 
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/3

  Issue #663:
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/663
	
	JC 2019-02-02 / Leilukin 2022-08-05                                             */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

	int nRomance;
  int nGlobal = GetGlobalNumber("K_SWG_CARTH");
	if (nGlobal > 12 && nGlobal < 16) nRomance = 0;
	else nRomance = 1;
	int nHKCheck = GetGlobalBoolean("Ebo_HK47");
	int nReturn;
	if (IsAvailableCreature(NPC_HK_47)) nReturn = nRomance * nHKCheck;
	else nReturn = nRomance;

	return nReturn;
}

**Comment by Leilukin** [comment]
Hi @Salk73 , thank you for your input. I compiled and tested the scripts you provided with a save that set the global of `K_SWG_CARTH` to 14, and unfortunately it did not enable Carth's romance-specific dialogue post-Leviathan.

For comparison, this video shows Carth's romance-specific dialogue with my scripts from the original post:
https://youtu.be/Jp0POoIzzz8

This video shows the scene with the updated K1CP script Salk provided above:
https://youtu.be/7EhMjugGu9w
I did not actually have the rest of K1CP installed in this particular save (bcauseI used many other mods in my K1CP saves), so Juhani did not show up in this video, but as shown in this video, the script still forces me to talk to HK-47.

**Comment by Salk73** [comment]
Hi Leilukin.

My apologies. When I read your message I thought you were suggesting combining the two fixes together so being forced to talk to HK47 first was actually intended. Now I understand that the dialogue tree probably needs re-arranging too. 

If DP doesn't beat me to it, I will take a look after the weekend.

**Comment by Salk73** [comment]
Hi again!

I took a closer look at the issue at hand and I have some remarks.

For starters, I wonder why entries 28 and 29 (E28 and E29) do not list reply 30 (R30). I don't see a reason for not including that even there.

Instead when it comes to  k_pebo_carthrom.ncs adn k_pebo_carthromm.ncs, it seems like JC's solution in the CP is going to destroy all of the Carth's romance lines as long as HK-47 is part of the team. And if that is the case, it is obviously unacceptable. 

We can still fix everything rather easily though by introducing two new scripts and adding them at the appropriate reply nodes (R30 and R32). These two are the original scripts which will contain Leilukin's fix.

```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	k_pebo_carthrom
	
	Carth romance check for post-Leviathan scene.
	
	The original script was not properly checking all states of the Carth 
  romance.

  Issue #663:
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/663
	
	Salk 2022-08-05                                             */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

  int nReturn = 0;
  int nRomCheck = GetGlobalNumber("K_SWG_CARTH");

  if (nRomCheck > 12 && nRomCheck < 16) {
    nReturn = 1;
  }

  return nReturn;
}
```
```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	k_pebo_carthromn
	
	Carth romance check for post-Leviathan scene.
	
	The original script was not properly checking all states of the Carth 
  romance.

  Issue #663:
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/663
	
	Salk 2022-08-05                                             */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

  int nReturn = 1;
  int nRomCheck = GetGlobalNumber("K_SWG_CARTH");

  if (nRomCheck > 12 && nRomCheck < 16) {
    nReturn = 0;
  }

  return nReturn;
}
```
And these two customized versions replacing the original conditional script at R30 and R32 will address issue #3:

```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	cp_pebo_carthrm
	
	Carth romance check for post-Leviathan scene.
	
	The conditions have been changed so the player will be forced to speak to
	HK-47 during this scene (if they have recruited him) so his storyline isn't
	left hanging.
	
	Returns true if the player has romanced Carth and they've talked to HK.
	
	Issue #3: 
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/3
	
	Salk 2022-08-05                                            */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

  int nReturn = 0;
  int nRomCheck = GetGlobalNumber("K_SWG_CARTH");

  int nJoleeCheck = GetGlobalBoolean("Ebo_Jolee");
  int nCandCheck = GetGlobalBoolean("Ebo_Canderous");
  int nHK47Check = GetGlobalBoolean("Ebo_HK47");
  int nT3M4Check = GetGlobalBoolean("Ebo_T3M4");

  if ((nJoleeCheck) && (nCandCheck) && (nHK47Check) && (nT3M4Check) && (nRomCheck > 12 && nRomCheck < 16)) {
      nReturn = 1;
  }

  return nReturn;
}
```
```
////////////////////////////////////////////////////////////////////////////////
/*	KOTOR Community Patch
	
	cp_pebo_carthrmn
	
	Carth romance check for post-Leviathan scene.
	
	The conditions have been changed so the player will be forced to speak to
	HK-47 during this scene (if they have recruited him) so his storyline isn't
	left hanging.
	
	Returns true if the player has NOT romanced Carth and they've talked to HK.
	
	Issue #3: 
	https://github.com/KOTORCommunityPatches/K1_Community_Patch/issues/3
	
	Salk 2022-08-05                                            */
////////////////////////////////////////////////////////////////////////////////

int StartingConditional() {

  int nReturn = 1;
  int nRomCheck = GetGlobalNumber("K_SWG_CARTH");

  int nJoleeCheck = GetGlobalBoolean("Ebo_Jolee");
  int nCandCheck = GetGlobalBoolean("Ebo_Canderous");
  int nHK47Check = GetGlobalBoolean("Ebo_HK47");
  int nT3M4Check = GetGlobalBoolean("Ebo_T3M4");

  if ((!nJoleeCheck) || (!nCandCheck) || (!nHK47Check) || (!nT3M4Check) || (nRomCheck > 12 && nRomCheck < 16)) {
      nReturn = 0;
  }

  return nReturn;
}
```
[ebo_carth.dlg]
...
EntryList\30\RepliesList\4\Active=cp_pebo_carthrmn
EntryList\30\RepliesList\5\Active=cp_pebo_carthrm
...

**Comment by JCarter426** [comment]
It seems like I overlooked that those scripts were used earlier in the conversation. For what it's worth, I don't think they break the romance, although I of course agree this should be fixed.

The code you posted, though, will make it impossible to continue the game if HK or Jolee (or theoretically T3 or Canderous) is not in the party. The existing code has an extra check for if HK wasn't recruited at all. I don't think checking for the other party members is needed, anyway.

I've updated my code with everyone's suggestions, and I'm about to commit. I hope that takes care of it this time.

**Comment by Leilukin** [comment]
Thank you both @Salk73 and @JCarter426 .

I discovered this issue when I was working on my Carth Onasi and Male PC Romance mod, and I have included my own fix for this scene in my Carth romance mod. It is true that missing out the Carth romance dialogue in this scene does not break Carth's romance, though I decided to bring up the issue here as I thought a global check issue like this is something K1CP would like to fix as well.

**Comment by Salk73** [comment]
Oh of course, JC. That was quite the oversight on my part up there.

Even if I am perfectly happy with your own amendment, I fixed my script too just for correctness's sake.

There is still the question about why we don't have R30 at E28 and E29 though. Locally I have added it at both nodes.

**Comment by DarthParametric** [comment]
I gather this was fixed per 404b9c1be8c39a391206d6ddef11d22a1b390b09 / ec736fa0ac8b33a01f5070c905f9bc4e89ffc9c5 ? Re-open if required.
